### PR TITLE
ENH: Add BDS test

### PR DESCRIPTION
**Pull request by ChadFulton**
Adds BDS test for IID time series, discussed in #920

### CODE REVIEW

**Comment by ChadFulton** [comment]
Rebased.

**Comment by ChadFulton** [comment]
Fixed the remaining issues. If it looks good, I will squash this down and merge.

File: statsmodels/tsa/bds.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+            ((embedding_dim - 1)**2) * (corrsum_1dim**(2 * embedding_dim)) -
+            (embedding_dim**2) * k * (corrsum_1dim**(2 * embedding_dim - 2)))
+
+    return variances, k
+
+
+def bds(x, max_dim=2, epsilon=None, distance=1.5):
+    """
+    Calculate the BDS test statistic for a time series
+
```
Comment (josef-pkt) [inline statsmodels/tsa/bds.py]: This needs more descriptive docstring, it doesn't say what BDS is or tests.
e.g. similar to the title of the article
"test for serial independence based on the correlation dimension"

Comment (josef-pkt) [inline statsmodels/tsa/bds.py]: also what's the null hypothesis, If pvalue is small, then we reject independence ?

**Comment by ChadFulton** [comment]
There's no need for so many commits here. I'll squash them this evening into a single commit unless there's some reason to keep them around.

**Comment by josef-pkt** [comment]
you still have triple quote docstrings in the unit tests.

I don't remember any changes that would make keeping separate commits necessary.

**Comment by ChadFulton** [comment]
> you still have triple quote docstrings in the unit tests.

I think that's okay because they're test classes and not test functions.

> I don't remember any changes that would make keeping separate commits necessary.

Yes, I think this just took a while because I had no idea what I was doing ;)

**Comment by ChadFulton** [comment]
Will merge in a few hours unless further comments.

**Comment by josef-pkt** [comment]
Especially given that we are trying for a release:

needs to be added to api and documentation.
Also, maybe the module should be privatised `_bds.py` and added to tsa/stattools for the official import/api path ?

otherwise no comment (looks fine), and I haven't read through the code again.

**Comment by ChadFulton** [comment]
Good idea, I'll privatize it and import it into `tsa/stattools`, and add to release notes.

**Comment by ChadFulton** [comment]
Privatized into _bds and put into stattools; added to documentation; squashed again; will merge if green.

**Pull request merged**